src/Components/Components/src/NavigationManager.cs

```diff
@@ -0,0 +1,220 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Components.Routing;
+
+namespace Microsoft.AspNetCore.Components
+{
+    /// <summary>
+    /// Provides an abstraction for querying and mananging URI navigation.
+    /// </summary>
+    public abstract class NavigationManager
+    {
+        /// <summary>
+        /// An event that fires when the navigation location has changed.
+        /// </summary>
+        public event EventHandler<LocationChangedEventArgs> LocationChanged
+        {
+            add
+            {
+                AssertInitialized();
+                _locationChanged += value;
+            }
+            remove
+            {
+                AssertInitialized();
+                _locationChanged -= value;
+            }
+        }
+
+        private EventHandler<LocationChangedEventArgs> _locationChanged;
+
+        // For the baseUri it's worth storing both the string form and Uri form and
+        // keeping them in sync. These are always represented as absolute URIs with
+        // a trailing slash.
+        private Uri _baseUri;
+        private string _baseUriString;
+
+        // The URI. Always represented an absolute URI.
+        private string _uri;
+
+        private bool _isInitialized;
+
+        /// <summary>
+        /// Gets or sets the current base URI. The <see cref="BaseUri" /> is always represented as an absolute URI in string form with trailing slash.
+        /// Typically this corresponds to the 'href' attribute on the document's &lt;base&gt; element.
+        /// </summary>
+        /// <remarks>
+        /// Setting <see cref="BaseUri" /> will not trigger the <see cref="LocationChanged" /> event.
+        /// </remarks>
+        public string BaseUri
+        {
+            get
+            {
+                AssertInitialized();
+                return _baseUriString;
+            }
+            protected set
+            {
+                if (value != null)
+                {
+                    value = NormalizeBaseUri(value);
+                }
+
+                _baseUriString = value ?? "/";
+                _baseUri = new Uri(_baseUriString);
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets the current URI. The <see cref="Uri" /> is always represented as an absolute URI in string form.
+        /// </summary>
+        /// <remarks>
+        /// Setting <see cref="Uri" /> will not trigger the <see cref="LocationChanged" /> event.
+        /// </remarks>
+        public string Uri
+        {
+            get
+            {
+                AssertInitialized();
+                return _uri;
+            }
+            protected set
+            {
+                _uri = value;
```

Comment:
Should this throw if you try to set a value that's not within the base URI space, or are you expanding the scope of this feature to be able to handle external URIs too?

Reply:
I'm happy to add some validation to these.